src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1312 @@
+        size_t mTotalCount        = 0;
+    };
+
+    class KeySetIteratorImpl : public KeySetIterator
+    {
+    public:
+        KeySetIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index) :
+            mProvider(provider), mFabric(fabric_index)
+        {
+            FabricStore fabric_store(fabric_index);
+            FabricInfo fabric_info;
+            if (CHIP_NO_ERROR == fabric_store.Load(fabric_info))
+            {
+                mNextId = fabric_info.first_keyset;
+                mCount  = fabric_info.keyset_count;
+                mIndex  = 0;
+            }
+        }
+
+        size_t Count() override { return mCount; }
+
+        bool Next(KeySet & outEntry) override
+        {
+            if (mIndex >= mCount)
+            {
+                return false;
+            }
+
+            KeySetInfo keyset_info;
+            if (CHIP_NO_ERROR != KeySetStore(mFabric, mNextId).Load(keyset_info))
+            {
+                return false;
+            }
+            mIndex++;
+            mNextId                = keyset_info.next;
+            outEntry.keyset_id     = keyset_info.keyset_id;
+            outEntry.policy        = keyset_info.policy;
+            outEntry.num_keys_used = keyset_info.num_keys_used;
+            memcpy(outEntry.epoch_keys, keyset_info.epoch_keys, sizeof(outEntry.epoch_keys));
+            return true;
+        }
+
+        void Release() override { mProvider.Release(this); }
+
+    private:
+        GroupDataProviderImpl & mProvider;
+        chip::FabricIndex mFabric = 0;
+        uint16_t mNextId          = 0;
+        size_t mCount             = 0;
+        size_t mIndex             = 0;
+    };
+
+public:
+    //
+    // Common
+    //
+
+    CHIP_ERROR Init() override
+    {
+#if CHIP_DEVICE_LAYER_TARGET_DARWIN
+        ReturnErrorOnFailure(DeviceLayer::PersistedStorage::KeyValueStoreMgrImpl().Init("chip.store"));
+#elif CHIP_DEVICE_LAYER_TARGET_LINUX
+        ReturnErrorOnFailure(DeviceLayer::PersistedStorage::KeyValueStoreMgrImpl().Init(CHIP_CONFIG_KVS_PATH));
+#endif
+        mInitialized = true;
+        return CHIP_NO_ERROR;
+    }
+
+    void Finish() override { mInitialized = false; }
+
+    //
+    // Group Mappings
+    //
+
+    bool HasGroupNamesSupport() override { return true; }
+
+    bool GroupMappingExists(chip::FabricIndex fabric_index, const GroupMapping & mapping) override
+    {
+        VerifyOrReturnError(mInitialized, false);
+
+        FabricStore fabric_store(fabric_index);
+        FabricInfo fabric_info;
+
+        CHIP_ERROR err = fabric_store.Load(fabric_info);
+        VerifyOrReturnError(CHIP_NO_ERROR == err, false);
+        VerifyOrReturnError(kInvalidEndpointId != fabric_info.first_endpoint, false);
+
+        // Existing fabric
+
+        chip::EndpointId endpoint_id = fabric_info.first_endpoint;
+        EndpointInfo endpoint_info;
+        size_t endpoint_count = 0;
+
+        // Loop through the fabric's endpoints
```

Comment:
The strategy I have implemented here is fabric pointing to a linked list of endpoints, each one pointing to a linked list of groups:
```
struct FabricData
{
    chip::EndpointId first_endpoint = kInvalidEndpointId;
    uint16_t endpoint_count         = 0;
    ...
};

struct EndpointData
{
    chip::GroupId first_group = kUndefinedGroupId;
    chip::EndpointId next     = kInvalidEndpointId;
    ...
};
```
As you mentioned, I chose this approach to make it closer to the requirements of the Groups cluster. However, a fabric->group->endpoint would work as well, since what we are implementing is logically a map<EndpointId, GroupId>. So as long as the endpoint-group pairs are unique, it doesn't matter if we store them as (endpoint, group), or (group, endpoint).
I don't think the order here has a performance impact. Nevertheless, we could optimize the overhead of the storage keys by saving a fixed-size list of endpoints (or groups) instead of each group (or endpoint) individually.